Exceptions. Refactoring.

ivr42 opened the pull request:
- added exception handler for `ApplicationError`
  - `ApplicationError` now inherited from `Exception`
  - child exception classes of `ApplicationError` with used response codes were added
  - all exceptions inherited from this child exceptions
  - all exception class names end with "Error"
  - In all modules, exceptions are imported as
    `from src.core import exceptions` and used as `exceptions.<ExceptionClass>`


Задача в [Notion](https://www.notion.so/Refactoring-5d04dc752032466db51d3880e90d916d)

ivr42 committed:
Exceptions. Refactoring. Stage 1

  - added exception handler for `ApplicationError`
  - `ApplicationError` now inherited from `Exception`
  - child exception classes of `ApplicationError` with used response codes
     were added
  - all exceptions inherited from this child exceptions

ivr42 committed:
ToDos were added

ivr42 committed:
Now all exception class names end with "Error"

ivr42 committed:
Even More Exception Refactoring

  - Got rid of 403 (Forbidden). Now using 400 (Bad Request) instead
  - Use models in `ObjectNotFoundError` and `ShiftStartError` exceptions
  - Use VK group address from settings
  - Use common `ShiftError` exception insead of:
    `CreateShiftForbiddenError`, `ShiftUpdateError`,
    and `UpdateShiftForbiddenError`

ivr42 committed:
Merge branch 'develop' into refactoring/exceptions

ivr42 committed:
Exceptions import and use refactored

  - In all modules, exceptions are imported as
    `from src.core import exceptions` and used as `exceptions.<ExceptionClass>`

ivr42 committed:
FORBIDDEN status code removed from router's responses

ADesBiysk commented on src/core/exceptions.py:
```diff
@@ -1,54 +1,50 @@
-        self.detail = "Объект {} с id: {} не найден".format(object_name, object_id)
+class ObjectNotFoundError(NotFoundError):
+    def __init__(self, model: DatabaseModel, object_id: UUID):
+        self.detail = "Объект '{}' с id '{}' не найден".format(model.__name__, object_id)
```
Предлагаю сюда передавать результат __repr__ модели. Это позволит:
1. избавиться от лишних манипуляций с данными
2. настраивать представление модели в логах и ошибках в одном месте.
3. Убрать проблемы с аннотацией типов и DatabaseModel.

ivr42 replied:
Тут у нас не получится так сделать (в других местах — да). Тут у нас нет **экземпляра**, только класс — объект же не найден, соответственно `model.__name__` даёт нам имя этого класса.

ADesBiysk commented on src/core/exceptions.py:
```diff
@@ -1,54 +1,50 @@
-        self.detail = f"Объект {obj} уже существует"
+class ObjectAlreadyExistsError(BadRequestError):
+    def __init__(self, model: DatabaseModel):
+        self.detail = "Объект {} уже существует".format(model)
```
Здесь также. И во всех исключениях, куда передаются данные о модели БД.

ADesBiysk commented on src/core/exceptions.py:
```diff
@@ -81,221 +77,144 @@ class ExceededAttemptsReportError(ApplicationError):
-        self.status_code = HTTPStatus.FORBIDDEN
-        self.detail = "Приглашение уже деактивировано"
+class InvitationAlreadyDeactivatedError(BadRequestError):
+    detail = "Приглашение уже деактивировано"
```
Предлагаю дополнить текст: Невозможно изменить состояние приглашения. Оно уже деактивировано.
И по аналогии ниже.

ADesBiysk commented on src/core/exceptions.py:
```diff
@@ -81,221 +77,144 @@ class ExceededAttemptsReportError(ApplicationError):
 
-class AdministratorBlockedException(ApplicationException):
+class AdministratorBlockedError(BadRequestError):
     """Попытка аутентификации заблокированного пользователя."""
```
Здесь и ниже, нужно уточнить, что речь про администратора, а не простого пользователя.

ADesBiysk commented on src/core/exceptions.py:
```diff
@@ -81,221 +77,144 @@ class ExceededAttemptsReportError(ApplicationError):
-    def __init__(self):
-        self.status_code = HTTPStatus.FORBIDDEN
-        self.detail = "У вас нет прав на изменение других администраторов."
+class AdministratorChangeError(BadRequestError):
```
А вот для последних 4 исключений статус Forbidden подошёл бы лучше. В других BadRequest нормально подходит.

ADesBiysk commented on src/core/services/shift_service.py:
```diff
@@ -64,8 +58,8 @@ def __init__(
         if _date <= _date.today():
-            raise ShiftUpdateException(
-                detail="Нельзя установить дату начала/окончания смены сегодняшним или прошедшим числом"
+            raise exceptions.ShiftError(
```
Это исключение получилось очень общим и используется в 7 случаях. По хорошему, это должны быть разные исключения.

ivr42 committed:
Code-review fixes. Stage 1.

  - Class `__repr__` used in appropriate error messages
  - Some error messages was changed
  - HTTP 403 (Forbidden) has been used for a few exceptions.

ivr42 committed:
Class `ShiftError` is split into seven detached classes.

ivr42 committed:
Some Report exceptions was refactored

ivr42 committed:
Merge branch `develop` into `refactoring/exceptions` and resolve the conflict
